bug fixes

remdragon opened the pull request:
avoid using signed integer index in for loop when terminating value is unsigned
check for malloc() returning NULL
detect overflow when converting BigInt to int
dedup BigInt_print()
BigInt_fprint() performance improvement

remdragon committed:
bug fixes

avoid using signed integer index in for loop when terminating value is unsigned
check for malloc() returning NULL
detect overflow when converting BigInt to int
dedup BigInt_print()
BigInt_fprint() performance improvement

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,108 @@
-    
+    if(!new_big_int) {
+        if(!errno) {
+            errno = ENOMEM;
```
From my reading of the [malloc man page](https://www.man7.org/linux/man-pages/man3/malloc.3.html#:~:text=On%20error%2C%20these%20functions%20return%20NULL%20and%20set%20errno.) it should always set errno on failure, so I don't think it's necessary to manage errno like this.

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,108 @@
-        value /= 10;
+    unsigned int count = new_big_int->num_digits;
+    char* digits = new_big_int->digits;
+    while( count-- ) {
```
```suggestion
    while(count--) {
```

dandclark commented on BigInt.h:
```diff
@@ -39,23 +45,24 @@ void BigInt_assign_int(BigInt* target, const int source);
+// value of big_int fits within the size of result's type on the target
+// environment.  returns non-zero on success or 0 on failure.
+// result is undefined on failure.
+BOOL BigInt_to_int(int* result, const BigInt* big_int);
```
```suggestion
BOOL BigInt_to_int(const BigInt* big_int, int* result);
```
This is nitpicky but I'd prefer to have the parameter order match the order of things like `BigInt_multiply_int` and match the fact that `BigInt` comes before `int` in the name of this function.

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,108 @@
-        new_big_int->digits[i] = value % 10;
-        value /= 10;
+    unsigned int count = new_big_int->num_digits;
+    char* digits = new_big_int->digits;
```
```suggestion
    unsigned char* digits = new_big_int->digits;
```
gcc emits a `-Wpointer-sign` warning otherwise.

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,108 @@
-    for(i = 0; i < source->num_digits; ++i) {
-        target->digits[i] = source->digits[i];
+    unsigned int count = source->num_digits;
+    char* tgt = target->digits;
```
```suggestion
    unsigned char* tgt = target->digits;
```

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,108 @@
-        target->digits[i] = source->digits[i];
+    unsigned int count = source->num_digits;
+    char* tgt = target->digits;
+    const char* src = source->digits;
```
```suggestion
    unsigned const char* src = source->digits;
```

dandclark commented on BigInt.c:
```diff
@@ -73,13 +112,17 @@ void BigInt_assign_int(BigInt* target, const int source) {
-    for(i = 0; i < target->num_digits; i++) {
-        target->digits[i] = value % 10;
+    unsigned int count = target->num_digits;
+    char* digits = target->digits;
```
```suggestion
    unsigned char* digits = target->digits;
```

dandclark commented on BigInt.c:
```diff
@@ -100,18 +143,22 @@ int BigInt_compare_digits(const BigInt* a, const BigInt* b) {
-    for(i = a->num_digits - 1; i >= 0; --i) {
-        if(a->digits[i] > b->digits[i]) {
+    unsigned int count = a->num_digits;
+    const char* pa = &a->digits[count-1];
```
```suggestion
    unsigned const char* pa = &a->digits[count-1];
```

dandclark commented on BigInt.c:
```diff
@@ -100,18 +143,22 @@ int BigInt_compare_digits(const BigInt* a, const BigInt* b) {
-        if(a->digits[i] > b->digits[i]) {
+    unsigned int count = a->num_digits;
+    const char* pa = &a->digits[count-1];
+    const char* pb = &b->digits[count-1];
```
```suggestion
    unsigned const char* pb = &b->digits[count-1];
```

dandclark commented on BigInt.c:
```diff
@@ -285,66 +367,101 @@ void BigInt_multiply(BigInt* big_int, const BigInt* multiplier) {
-        tens_multiplier *= 10;
-    } 
+    unsigned int num_digits = big_int->num_digits;
+    const char* digits = big_int->digits;
```
```suggestion
    unsigned const char* digits = big_int->digits;
```

dandclark commented on BigInt.c:
```diff
@@ -285,66 +367,101 @@ void BigInt_multiply(BigInt* big_int, const BigInt* multiplier) {
-    if (big_int->is_negative) fprintf(dest, "-");
-    for(i = big_int->num_digits - 1; i >= 0; --i) {
-        fprintf(dest, "%i", big_int->digits[i]);
+    const char* base = big_int->digits;
```
```suggestion
    unsigned const char* base = big_int->digits;
```

dandclark commented on BigInt.h:
```diff
@@ -26,10 +32,10 @@ void BigInt_free(BigInt* big_int);
 ///Sets the value of the target BigInt to the value of the source BigInt.
 // Assumes that target and source already point to valid BigInts. 
-void BigInt_assign(BigInt* target, const BigInt* source);
+BOOL BigInt_assign(BigInt* target, const BigInt* source);
```
Would be helpful to update the header comments to say what the return value means for all these.

remdragon replied:
Are you looking for "returns non-zero on success or 0 on failure", or are you looking something more like a list of possible reasons for the function to fail? I'm going with the former for now, but want to make sure I understand what you're requesting completely (BOOL implies "returns non-zero on success or 0 on failure" to me, which is why I'm unsure exactly what's being asked for here, sorry)

dandclark replied:
What you've got now ("returns non-zero on success or 0 on failure") is good.

dandclark reviewed (changes requested):
Thanks for the PR!

Left some comments from a first review pass. `make` is failing to build with clang because demo.c needs to be updated to handle the `BigInt_to_int` change, and there are some `unsigned char*` to `char*` conversions that yield warnings -- I left comments for those. Can you take a look?

remdragon committed:
clang compatibility, fix buffer misallocation

fix clang compatibility issues (char* vs unsigned char*)
add optional redzone logic to help find seg fault
fix bug in BigInt_ensure_digits() where not enough memory was allocated (causing seg fault)
comment out BigInt_test_operations(999999999,123456789) because it always fails with ERANGE trying to convert back to int
rename OPERATION_TYPE_MAX to OPERATION_TYPE_COUNT because MAX implies maximum value whereas COUNT implies maximum value + 1
allow setting BIGINT_TEST_LOGGING outside of BigInt_test.h
print "testing finished" at the end of all tests (tinyc does not print anything when a seg fault occurs)

remdragon commented:
I was stupidly only testing with tinyc. I'm now testing with clang and gcc as well.

remdragon committed:
changes requested by danddclark

reorder args in BigInt_to_int()
document return values for functions that now return BOOL instead of void
fix demo.c to be compatible with api changes
fix README.md to be compatible with api changes

dandclark commented on BigInt_test.c:
```diff
@@ -63,7 +67,7 @@ void BigInt_test_basic() {
     BigInt_test_operations(2546, 2546);
     BigInt_test_operations(1234, 4321);
-    BigInt_test_operations(999999999, 123456789);
+    //BigInt_test_operations(999999999, 123456789);
```
```suggestion
    BigInt_test_operations(999999999, 123456789);
```
Looks like this last test still fails. I haven't had time to dig into it further yet but probably best to understand it before merging.

```
Testing basic 2-argument operations
Assertion failed: (BigInt_to_int(big_int_a, &result)), function BigInt_test_single_operation, file BigInt_test.c, line 172.
```

remdragon replied:
The test is failing because it can't be converted back to int. 999999999 * 123456789 doesn't fit into a 32-bit int.

dandclark replied:
Ah, I see.

dandclark commented on BigInt_test.c:
```diff
@@ -29,11 +31,13 @@ void BigInt_test_basic() {
+    int value;
+    assert(BigInt_to_int(big_int, &value) && value == 42);
+    assert(BigInt_ensure_digits(big_int, 1000));
+    assert(BigInt_to_int(big_int, &value) && value  == 42);
```
```suggestion
    assert(BigInt_to_int(big_int, &value) && value == 42);
```

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,157 @@
+    new_big_int->digits = malloc_digits(new_big_int->num_allocated_digits);
+    if(!new_big_int->digits) {
+        free(new_big_int);
+        if(!errno) {
```
Same feedback as the other comment, it seems unnecessary to manage `errno` for `malloc` like this (or from `malloc_digits` since that's controlled in this code).

dandclark commented on BigInt.c:
```diff
@@ -285,66 +416,103 @@ void BigInt_multiply(BigInt* big_int, const BigInt* multiplier) {
+    
+    // don't process any most significant digits that happen to be zero
+    // (avoids unnecessary tens_multiplier overflow)
+    while( num_digits && !digits[num_digits-1] ){
```
```suggestion
    while(num_digits && !digits[num_digits-1]){
```

dandclark commented on README.md:
````diff
@@ -29,9 +29,10 @@ The contents of a BigInt can be printed to stdout with a call to BigInt_print.
 ```
 
-Get a normal int back from a BigInt with using BigInt_to_int; but be careful; the BigInt must fit in an int type!
+Get a normal int back from a BigInt with using BigInt_to_int; but be careful; if the BigInt doesn't fit in an int type, the function will return 0 and the output will be unchanged and 
````
```suggestion
Get a normal int back from a BigInt with using BigInt_to_int; but be careful; if the BigInt doesn't fit in an int type, the function will return 0 and the value in the int argument will be undefined.
```

Looks like BigInt_to_int can trample on the out param prior to hitting an overflow condition, so probably safest to just say the output is undefined on failure.

dandclark commented on BigInt.c:
```diff
@@ -285,66 +416,103 @@ void BigInt_multiply(BigInt* big_int, const BigInt* multiplier) {
-        fprintf(dest, "%i", big_int->digits[i]);
+    const unsigned char* base = big_int->digits;
+    const unsigned char* digits = &base[big_int->num_digits-1];
+    if (big_int->is_negative) fputc('0', dest);
```
```suggestion
    if (big_int->is_negative) fputc('-', dest);
```

dandclark commented on BigInt.c:
```diff
@@ -1,69 +1,157 @@
+    return p + BIGINT_REDZONE;
+}
+
+unsigned char* okay_digits(unsigned char* digits, unsigned int num_digits) {
```
This is clever. Maybe worth a brief comment at the top of the redzone block outlining the checking scheme?

dandclark reviewed (changes requested):
Thanks for the follow-ups, this is close to ready to merge.
Just had some remaining comments for: a failing test, a small bug in BigInt_fprintf, and a few misc nitpicks.

remdragon committed:
documentation, formatting, less errno handling, bug fix

document redzone logic
apply formatting requested by dandclark
don't manipulate errno unnecessarily
fix bug outputting wrong symbol for negative numbers in BigInt_fprint()
add asserts to demo.c
fix inaccuracy in README.md

dandclark reviewed (approved):
Thanks again for the PR!
